torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2357,103 +2357,325 @@ Tensor slice_backward_wrapper(
+    return ret;
+  }();
+
+  const auto sym = [](const Tensor& X) { return X + X.mH(); };
+
+  // Im(diag(dP)) / (2S)
+  auto ImdiagdP2S = is_complex ? at::imag(dP.diagonal(0, -2, -1)).div(2. * S)
+                               : Tensor{};
+
+  // dU = U (sym(dP S) / E) + i Im(diag(dP)) / (2S)
+  auto dU = [&] {
+    auto dUaux = sym(dP * S.unsqueeze(-2)) / E;
+    if (is_complex) {
+      at::imag(dUaux.diagonal(0, -2, -1)).copy_(ImdiagdP2S);
     }
-    if (gv.defined()) {
-      gv = gv.narrow(-1, 0, k);
+    return at::matmul(U, dUaux);
+  }();
+  if (m > n) {
+    // dU += (I_m - UU^H) dA V S^{-1}
+    const auto dAVSinv = at::matmul(dA, V / S.unsqueeze(-2)) ;
+    dU += dAVSinv - at::matmul(U, at::matmul(U.mH(), dAVSinv));
+
+    // To "fix" the full_matrices case (the full_matrices case should not be differentiable...)
+    if (full_matrices) {
+      auto shape = dU.sizes().vec();
+      shape.end()[-1] = m - n;
+      dU = at::cat({dU, at::zeros(shape, dU.options())}, /*dim=*/-1);
     }
   }
-  auto vh = v.mH();
 
-  Tensor sigma_term;
-  if (gsigma.defined()) {
-    gsigma = gsigma.to(U.dtype());
-    // computes u @ diag(gsigma) @ vh
-    sigma_term = at::matmul(u * gsigma.unsqueeze(-2), vh);
-  } else {
-    sigma_term = at::zeros_like(m >= n ? u : vh, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
-  }
-  // in case that there are no gu and gv, we can avoid the series of kernel
-  // calls below
-  if (!gv.defined() && !gu.defined()) {
-    return sigma_term;
-  }
-
-  auto uh = u.mH();
-  auto sigma_inv = sigma.pow(-1);
-  auto sigma_sq = sigma.pow(2);
-  auto F = sigma_sq.unsqueeze(-2) - sigma_sq.unsqueeze(-1);
-  // The following two lines invert values of F, and fills the diagonal with 0s.
-  // Notice that F currently has 0s on diagonal. So we fill diagonal with +inf
-  // first to prevent nan from appearing in backward of this function.
-  F.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).fill_(INFINITY);
-  F = F.pow(-1);
-
-  Tensor u_term, v_term;
-
-  if (gu.defined()) {
-    auto guh = gu.mH();
-    u_term = at::matmul(u, F.mul(at::matmul(uh, gu) - at::matmul(guh, u)) * sigma.unsqueeze(-2));
-    if (m > k) {
-      // projection operator onto subspace orthogonal to span(U) defined as I - UU^H
-      auto proj_on_ortho_u = -at::matmul(u, uh);
-      proj_on_ortho_u.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).add_(1);
-      u_term = u_term + proj_on_ortho_u.matmul(gu * sigma_inv.unsqueeze(-2));
+  // dVh = -sym(S dP) / E + i Im(diag(dP)) / (2S)
+  // Perf: We negate the S as it's the smallest tensor in the equation
+  auto dVh = [&] {
+    auto dVhaux = sym(dP * (-S).unsqueeze(-1)) / E;
+    if (is_complex) {
+      at::imag(dVhaux.diagonal(0, -2, -1)).copy_(ImdiagdP2S);
+    }
+    return at::matmul(dVhaux, Vh);
+  }();
+  if (m < n) {
+    // dVh += S^{-1} U^H dA (I_n - VV^H)
+    const auto UHdASinv = at::matmul(U.mH() / S.unsqueeze(-1), dA) ;
+    dVh += UHdASinv - at::matmul(at::matmul(UHdASinv, V), Vh);
```

Comment:
perf nit: maybe we can write a function that makes these products that reduce over the largest dimension and do not produce `max(m, n) x max(m, n)` matrices.

Reply:
In fact, we can write a utility function that performs these kind of projections, as this one is quite heavily used in the backward context.

Reply:
So, if I'm not mistaken, this never instantiates a large matrix. Note that `V` is of size `(n, m)` with `n > m`. As such, doing `UHdASinv @ V` gives a small matrix of size `(n, n)`. Then this matrix is mapped back onto the large space by multiplying by `Vh`. The size of the largest matrix is the size of the input `dA`.

I tried to be careful in this sense when writing all this code. Please tell me if you find any place where this does not hold, but I believe it holds here.

Reply:
Exactly, `n > m`, but a matrix `n x n` is produced.

Reply:
When? I don't think so...

Reply:
sorry, it should said `(m, m)` over there. Editing

Reply:
It just follow from your comment.

Reply:
No `(n, n)` matirx is computed here. This is the correct way of associating things when dealing with a projection. Note that `V` is a tall matrix and `A` is a wide matrix, so the result is a small matrix

Reply:
All good, sorry, I forgot that U is `(m, m)`.